diff -uraN fast-usbserial-master/Descriptors.c fast-usbserial/Descriptors.c
--- fast-usbserial-master/Descriptors.c	2015-10-24 22:36:00.000000000 +0300
+++ fast-usbserial/Descriptors.c	2021-09-15 21:40:19.327300111 +0300
@@ -72,7 +72,7 @@
 
 	.ManufacturerStrIndex   = 0x01,
 	.ProductStrIndex        = 0x02,
-	.SerialNumStrIndex      = USE_INTERNAL_SERIAL,
+	.SerialNumStrIndex      = 0x03,
 
 	.NumberOfConfigurations = FIXED_NUM_CONFIGURATIONS
 };
@@ -226,6 +226,16 @@
 
 };
 
+const USB_Descriptor_String_t PROGMEM SerialString =
+{
+                .Header                 = {.Size = USB_STRING_LEN(5), .Type = DTYPE_String},
+
+                .UnicodeString          = L"PROG1"  //<---- ATTENTION
+                                                    //Uniq serial for device, don`t forget to correct USB_STRING_LEN !!!
+
+};
+
+
 /** This function is called by the library when in device mode, and must be overridden (see library "USB Descriptors"
  *  documentation) by the application code so that the address and size of a requested descriptor can be given
  *  to the USB library. When the device receives a Get Descriptor request on the control endpoint, this function
@@ -267,6 +277,10 @@
 					Address = (void*)&ProductString;
 					Size    = pgm_read_byte(&ProductString.Header.Size);
 					break;
+                                case 0x03:
+                                        Address = (void*)&SerialString;
+                                        Size    = pgm_read_byte(&SerialString.Header.Size);
+                                        break;
 			}
 
 			break;
